### ISSUE TITLE
c8d: `buildx build`, `load`, and `import` don't preserve dangling image

### ISSUE THREAD

**Issue by yosifkit** [moby/moby#48907]
### Description

With `containerd` enabled, `docker buildx build -t repo:tag` (whether using the in-daemon build or a `docker-container` builder), `docker import` and `docker load` fail to preserve an image that already exists as the specified `repo:tag`.

`docker pull` and `docker build` (classic) do not exhibit this failure.

### Reproduce

```console
$ docker pull busybox:latest
Using default tag: latest
latest: Pulling from library/busybox
430378704d12: Download complete
Digest: sha256:5b0f33c83a97f5f7d12698df6732098b0cdb860d377f6307b68efe2c6821296f
Status: Downloaded newer image for busybox:latest
docker.io/library/busybox:latest
$ docker images
...
busybox                                   latest          5b0f33c83a97   7 weeks ago     6.56MB

$ docker buildx build --progress=plain -t busybox:latest - <<EOF
FROM debian:bullseye
RUN touch foo
EOF
...
#6 exporting manifest sha256:3e76f5ea99fc0e229db1d62904b35049898a3bf08bbc0ca1038ad72604361926 done
#6 exporting config sha256:0962991c94f182518173b0e367ffe896a2609643b1fc8a9d78891224a06355b0 done
#6 exporting attestation manifest sha256:af89643f4c323de8d5a1ffb48df44109e30f547fc76e1144befee5702874df2a 0.0s done
#6 exporting manifest list sha256:ff5864949de931b93494acac119c99f4f7e6cb001860fe0486ac41f222744896
#6 exporting manifest list sha256:ff5864949de931b93494acac119c99f4f7e6cb001860fe0486ac41f222744896 done
#6 naming to docker.io/library/busybox:latest done
#6 unpacking to docker.io/library/busybox:latest 0.0s done
#6 DONE 0.1s
$ docker images
REPOSITORY                                TAG             IMAGE ID       CREATED          SIZE
busybox                                   latest          ff5864949de9   17 seconds ago   197MB
...
$ # 5b0f33c83a97 is not is the images output
$ $ docker image inspect 5b0f33c83a97
[]
Error response from daemon: No such image: 5b0f33c83a97:latest
```
The same happens if you replace the `docker buildx build ...` with a `docker load` that contains an image with that tag or a `docker import /dev/null busybox:latest`.

### Expected behavior

`docker buildx build`, `docker load` and `docker import` should behave the same way as `docker pull` and `docker build` (classic) and keep the old image as a dangling reference.

### docker version

```yaml
Client: Docker Engine - Community
 Version:           27.3.1
 API version:       1.47
 Go version:        go1.22.7
 Git commit:        ce12230
 Built:             Fri Sep 20 11:41:11 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Desktop  ()
 Engine:
  Version:          27.3.1
  API version:      1.47 (minimum version 1.24)
  Go version:       go1.22.7
  Git commit:       41ca978
  Built:            Fri Sep 20 11:41:11 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.7.21
  GitCommit:        472731909fa34bd7bc9c087e4c27943f9835f111
 runc:
  Version:          1.1.13
  GitCommit:        v1.1.13-0-g58aa920
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```yml
Client: Docker Engine - Community
 Version:    27.3.1
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.17.1-desktop.1
    Path:     /usr/local/lib/docker/cli-plugins/docker-buildx

... removed irrelevant plugins ...

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 30
 Server Version: 27.3.1
 Storage Driver: overlayfs
  driver-type: io.containerd.snapshotter.v1
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 nvidia runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 472731909fa34bd7bc9c087e4c27943f9835f111
 runc version: v1.1.13-0-g58aa920
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
 Kernel Version: 5.15.153.1-microsoft-standard-WSL2
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 20
 Total Memory: 31.18GiB
 Name: docker-desktop
 ID: 78016afb-38ee-4a54-a6ca-19d78c22412c
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Labels:
  com.docker.desktop.address=unix:///var/run/docker-cli.sock
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
```

### Additional Info

Loading in non-containerd would warn users that the image would be renamed as the "empty string" https://github.com/moby/moby/blob/1c00dfc77e115977d9127371d1d1df4a0878cb58/image/tarexport/load.go#L248

**Comment by jsternberg** [comment]
Import appears to be working after #48316. Load doesn't seem to work and I've opened https://github.com/moby/moby/pull/49650 to address that. The `buildx build` portion of this uses another different path for creating the image in the store and will require a different code change.

**Comment by jsternberg** [comment]
I've been doing a deep dive into this and I think it might help to discuss how dangling images are handled. The current code seems to try and find the places where we can overwrite an image and ensure the old image is marked as dangling before we overwrite it. This creates a small area where we have an inconsistent state that an image is both tagged with its normal name and also tagged as "dangling" in containerd. If something fails in between these two points (ensuring the dangling image and overwriting the old image with the new one), we have a state in the system that's not expected or needs to be cleaned up in some way.

At the same time, it seems like `docker images` already filters out the duplicate dangling entries from containerd. If we modify `pull` to create both the dangling reference and the named reference (see below `ctr i ls` output), the `docker images -a` output only prints this image once and with the correct name.

```
root@b5e069e8601d:/go/src/github.com/docker/docker# ctr i ls
REF                                                                                   TYPE                                    DIGEST                                                                  SIZE    PLATFORMS                                                                                                                          LABELS
docker.io/library/busybox:latest                                                      application/vnd.oci.image.index.v1+json sha256:498a000f370d8c37927118ed80afe8adc38d1edcbfc071627d17b25c88efcab0 1.8 MiB linux/386,linux/amd64,linux/arm/v5,linux/arm/v6,linux/arm/v7,linux/arm64/v8,linux/mips64le,linux/ppc64le,linux/riscv64,linux/s390x -
moby-dangling@sha256:498a000f370d8c37927118ed80afe8adc38d1edcbfc071627d17b25c88efcab0 application/vnd.oci.image.index.v1+json sha256:498a000f370d8c37927118ed80afe8adc38d1edcbfc071627d17b25c88efcab0 1.8 MiB linux/386,linux/amd64,linux/arm/v5,linux/arm/v6,linux/arm/v7,linux/arm64/v8,linux/mips64le,linux/ppc64le,linux/riscv64,linux/s390x -
```

docker images output

```
REPOSITORY   TAG       IMAGE ID       CREATED        SIZE
busybox      latest    498a000f370d   5 months ago   6.02MB
```

Removing the `busybox` image with `rmi` also seems to remove the dangling reference too.

Since this functionality already works, it would be decently easy to modify `pull`, `import`, `load`, and `build` to always tag newly created images with the dangling tag. We could keep the previous logic just in case there's an inconsistency, but the default expected state is that these images would have a dangling tag on them. With the above change to `pull`, all of the other commands get fixed when they overwrite the image created by pull.

If that change is ok, I can make it to these commands.

**Comment by thaJeztah** [comment]
This is all something we need to look at closely though; the whole reason we implemented this hack was to preserve the old behavior, which effectively meant: no content would be removed, ever, unless the user explicitly removed it (remove the image or `docker system prune` / `docker image prune`).

But effectively we're doing a lot of work to work around features that containerd provides; containerd can handle garbage collection, which would mean the user would no longer _have to_ manually micro-manage the content. Ultimately, that's the direction we should go into, but there were still some situations where that was possibly too disruptive; also because we don't have proper mechanisms to track the origin of an image:

- for an image that's pulled from a registry, we should assume that the registry is the source of truth; the local image is only a 'cached' copy of that image. Such images we could delete without much concern, because the content is not removed permanently (just the cached copy is removed).
- for an image that was built locally, but never pushed, things are more "risky". The local image _IS_ the source of truth, and removing the image would permanently destroy the image data (with no way to recover (e.g. by pulling it again)).

I think even for the last variant, it'd still be ok to remove the content after some time; an image should have a tag if it's important enough to preserve; if there's no tag, then it should be open for garbage-collection. If we move towards that, we should have some (possibly configurable) grace period before garbage collecting; i.e., make sure that images that were produced from repeated builds (so newer builds "untagging" the previous build) do not immediately garbage-collect the previous builds.

:point_up: I think that's the scenario that was reported; i.e., a `docker build` that produces a new version of `image:tag` would "untag" the previous build, but that image is also gone immediately (or shortly after) after that.

**Comment by thaJeztah** [comment]
FWIW; I have a draft PR open on docker/cli to move towards hiding untagged content by default. That PR does not change any of the behavior on how we manage the content, but is mostly a stepping-stone towards a future where we more clearly indicate "images without a name should be considered removed, or candidate for garbage-collection"; hiding the content by default is closer to that model.


- https://github.com/docker/cli/pull/5924

**Comment by thaJeztah** [comment]
cc @dmcgowan @vvoland - this somewhat relates to  discussions we had recently as well. ISTR @dmcgowan also mentioned options for images to be garbage-collected based on "parent" image they're part of and/or containerd now supporting a grace-period (to prevent _immediate_ garbage collection?)

**Comment by jsternberg** [comment]
This should be working now that #49702 is merged.

**Comment by jsternberg** [comment]
@thaJeztah is this issue one that we want to pursue on fixing? Now that containerd is the default, it seems likely more worth it to document the difference in this behavior and then move towards utilizing the garbage collection in containerd. I can try to put together a change that removes the dangling image logic just to see what that might look like.

**Comment by thompson-shaun** [comment]
@jsternberg - let's close this and not replicate for c8d and open a separate issue to discuss how to represent this and how it'll be handled as GC gets rolled out

**Comment by jsternberg** [comment]
I'll follow up with trying to understand how the GC stuff works but let's close this for now. For the record, I believe `load` and `import` preserve dangling images right now. `build` is the only one that doesn't.
